File: leetcode solutions/C++/Reverse Linked List/reverseLL.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x),next(NULL) {}
};

// Iterative
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL;

        while(head) {
            ListNode* next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};


// Recursive
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if ((!head ) || !(head->next)) return head;

        ListNode* node = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return node;
    }
};

File: leetcode solutions/Python/Reverse Linked List/reverseLL.py
class ListNode(object):
    def __init__(self):
        self.val = x
        self.next = None


# Iterative

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        prev = None
        curr = head

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        return prev



# recursive

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        return self.reverseListHelper(head, None)

    def reverseListHelper(self, node, prev=None):
        if not node:
            return prev

        nextNode = node.next
        node.next = prev
        return self.reverseListHelper(nextNode, node)

